Feature sets

tanema opened the pull request:
fixes #620

`shopify config feature` is a hidden command that would allow us to disable some features by default, for instance, Shopify plus features that most developers do not have access to. It has the added benefit of allowing teams to enable their hidden projects and have the CLI operate like they are no longer hidden.

This will allow for checking if a feature is enabled `ShopifyCli::Feature.enabled?(:shopify_plus)` in code paths and will also conditionally show hidden project types and hidden commands if the correct features are enabled

**Example: enabling a debug feature set and it shows the load-dev, load-system, and feature commands**
![image](https://user-images.githubusercontent.com/463193/87337721-e4163d00-c511-11ea-86f7-c1a8e76d96da.png)

attaboy commented:
In keeping with our preference for verbs over nouns, and also thinking about extensibility, I wonder about using `configure` for this, e.g. `shopify configure feature`, and then it could even potentially prompt you for the feature you want to enable/disable?

gfscott commented:
Very cool — I came here to say what Luke said!

`shopify config feature <feature_name> --enable` 
`shopify config feature <feature_name> --disable`

ErinRenNumber1 commented:
Just to clarify, this feature will allow enabling/disabling features at 2 levels?
1. project type level
2. commands within a project type
`shopify_plus` is probably just an example you used, but it's not a concept for Partners, and only a concept for merchants.

So, essentially, we could flexibly hide/unhide any project type, which might be very useful, if one day CLI supports 20 different project types, and as a developer, I only ever work with 2 such types.

mkevinosullivan commented:
This is a very neat idea!

Could there ever be a case where a command could belong to multiple feature sets, or is that just getting to far ahead of the current need?  (That is, deal with it if it's needed in the future.)

tanema commented:
Rebooting this. I updated the description. It now has 
- tests
- multiple feature flags for the `hidden_feature` method so that multiple flags can be used
- changed the command to use `shopify config feature [feature_name]` once for enable, again for disable

attaboy commented:
> changed the command to use shopify config feature [feature_name] once for enable, again for disable

I think it would be nice to have a command to force the feature on or off regardless of its current state, so you don't have to waste time if you're not sure whether you previously turned it on. I like Graham's idea of `--enable` and `--disable` (and perhaps we could have `--status` to just show you whether it's enabled).

Beyond that, this seems solid!

andyw8 commented on lib/shopify-cli/feature.rb:
```diff
@@ -0,0 +1,97 @@
+      #     ShopifyCli::Commands::Config.hidden?
+      #
+      def hidden?
+        enabled = (@hidden_feature_set || []).any? do |feature|
```
Is `|| []` needed? (Can `@hidden_feature_set` ever be falsey?)

tanema replied:
it can be nil.

andyw8 replied:
Ah, if `hidden_feature` is never called.

tanema commented:
Update to usability.

![image](https://user-images.githubusercontent.com/463193/87467986-d1ba0300-c5e6-11ea-9456-9a07b75773c6.png)

ErinRenNumber1 commented on lib/project_types/script/cli.rb:
```diff
@@ -2,7 +2,7 @@
 module Script
   class Project < ShopifyCli::ProjectType
-    hidden_project_type
+    hidden_feature
```
Should this be `hidden_feature(feature_set: :script)`?

tanema replied:
I left this hidden so that your team could decide on the naming of your feature set.

mkevinosullivan commented on lib/shopify-cli/extension.rb:
```diff
@@ -0,0 +1,4 @@
+module ShopifyCli
+  class Extension
```
This doesn't seem to be used anywhere?

tanema replied:
Oh you are right, this is actualy a miss-commit and part of other experimentation that I was doing. Thanks for catching that!

tanema commented:
re: Enable/Disable
You are enabling the visibility. Also the feature set can be used for much more than visiblity. It can be used to turn on/off code paths as well.

re: --list
This doesnt keep a list of registered feature flags and I do not want to enable easy discoverability. This is used for functionality that we may not want every user to use.

andyw8 commented:
We're missing some clean-up for the tests added here: https://github.com/Shopify/shopify-app-cli/issues/915
